lib/improver/nbhood.py

```diff
@@ -200,26 +160,125 @@ def _get_grid_x_y_kernel_ranges(self, cube, radius_in_km):
+        are in the nth row, and then cumulating along the x direction,
+        so that the largest values are in the mth column. Each grid point
+        will contain the cumulative sum from the origin to that grid point.
 
         Parameters
         ----------
         cube : Iris.cube.Cube
-            Cube containing the x and y coordinates, which will be used for
-            calculating the number of grid cells in the x and y direction,
-            which equates to the size of the desired radii.
+            Cube to which the cumulative summing along the y and x direction
+            will be applied.
+
+        Returns
+        -------
+        cube : Iris.cube.Cube
+            Cube to which the cumulative summing along the y and x direction
+            has been applied.
+
+        """
+        yname = cube.coord(axis="y").name()
+        xname = cube.coord(axis="x").name()
+        cubelist = iris.cube.CubeList([])
+        for slice_2d in cube.slices([yname, xname]):
+            data = slice_2d.data
+            data_summed_along_y = np.cumsum(data, axis=0)
+            data_summed_along_x = (
+                np.cumsum(data_summed_along_y, axis=1))
+            slice_2d.data = data_summed_along_x
+            cubelist.append(slice_2d)
+        return cubelist.merge_cube()
+
+    @staticmethod
+    def run(cube):
+        """
+        Call the methods required to apply a square neighbourhood
+        method to a cube.
+
+        Parameters
+        ----------
+        cube : Iris.cube.Cube
+            Cube containing the array to which the square neighbourhood
+            will be applied
+
+        Returns
+        -------
+        cube : Iris.cube.Cube
+            Cube containing the smoothed field after the square neighbourhood
+            method has been applied.
+        """
+        summed_up_cube = SquareNeighbourhood.cumulate_array(cube)
+        return summed_up_cube
+
+
+class CircularNeighbourhood(object):
+
+    """
+    Methods for use in the calculation and application of a circular
+    neighbourhood.
+
+    A maximum kernel radius of 500 grid cells is imposed in order to
+    avoid computational ineffiency and possible memory errors.
+    """
+
+    # Maximum radius of the neighbourhood width in grid cells.
+    MAX_RADIUS_IN_GRID_CELLS = 500
+
+    def __init__(self, radius_in_km, unweighted_mode=False):
+        """
+        Initialise class.
+
+        Parameters
+        ----------
+        radius_in_km : Float
+            Radius in kilometres for use in specifying the number of
+            grid cells used to create a circular neighbourhood.
+        unweighted_mode : boolean
+            If True, use a circle with constant weighting.
+            If False, use a circle for neighbourhood kernel with
+            weighting decreasing with radius.
+
+        """
+        self.radius_in_km = radius_in_km
+        self.unweighted_mode = unweighted_mode
+
```

Comment:
I've done something like this, by defining a variable called `self.neighboourhood_method_name`, so that I can access this within the `__repr__` method.